components/daikin_s21/s21.cpp

```diff
@@ -192,224 +147,476 @@ std::string str_repr(uint8_t *bytes, size_t len) {
+      switch (rcode[1]) {
+        case '1':
+          this->activate_climate = false;
+          break;
+        case '5':
+          this->activate_swing_mode = false;
+          break;
+        default:
+          break;
+      }
+      this->refresh_state = true; // a command took, trigger immediate refresh
+      return;
+
+    default:
+      break;
   }
-  ESP_LOGD(TAG, "Unknown response %s -> \"%s\"", str_repr(rcode).c_str(),
-           str_repr(payload).c_str());
-  return false;
-}
 
-bool DaikinS21::run_queries(std::vector<std::string> queries) {
-  bool success = true;
+  // protocol decoding debug
+  // note: well known responses return directly from the switch statements
+  // break instead if you want to view their contents
 
-  for (auto q : queries) {
-    std::vector<uint8_t> code(q.begin(), q.end());
-    success = this->s21_query(code) && success;
+  // print everything
+  // if (this->debug_protocol) {
+  //   ESP_LOGD(TAG, "S21: %s -> %s %s", rcode,
+  //            str_repr(payload, payload_len).c_str(),
+  //            hex_repr(payload, payload_len).c_str());
+  // }
+
+  // print changed values
+  if (this->debug_protocol) {
+    auto curr = std::vector<uint8_t>(payload, payload + payload_len);
+    if (val_cache[rcode] != curr) {
+      const auto prev = val_cache[rcode];
+      ESP_LOGI(TAG, "S21 %s changed: %s %s -> %s %s", rcode,
+               str_repr(prev.data(), prev.size()).c_str(),
+               hex_repr(prev.data(), prev.size()).c_str(),
+               str_repr(curr.data(), curr.size()).c_str(),
+               hex_repr(curr.data(), curr.size()).c_str());
+      val_cache[rcode] = curr;
+    }
+  }
+}
+
+void DaikinS21::handle_nak() {
+  ESP_LOGW(TAG, "Rx: NAK from S21 for %s", tx_command);
+  if (strcmp(tx_command, *current_query) == 0) {
+    ESP_LOGW(TAG, "Removing %s from query pool (assuming unsupported)", tx_command);
+    // current_query iterator will be invalidated, recover index and 
+    const auto index = std::distance(queries.begin(), current_query);
+    queries.erase(current_query);
+    current_query = queries.begin() + index;
+  } else {
+    ESP_LOGW(TAG, "Acknowledging %s command despite NAK", tx_command);
+    parse_ack();  // don't get stuck retrying unsupported command
   }
+}
+
+void DaikinS21::setup() {
+  // populate messages to poll
+  // clang-format off
+#define S21_EXPERIMENTS 1
+  queries = {
+      "F1", "F5", "F9",
+      "Rd", "RH", "RI", "Ra", "RL", "RG",
+      // redundant/worse: "RC", "RF", "RB",
+#if S21_EXPERIMENTS
+      // Observed BRP device querying these.
+      // "F2", "F3", "F4", "RN",
+      // "RX", "RD", "M", "FU0F",
+      // Query Experiments
+      // "RA", 
+      // "RE",
+      // "RK", "RM", "RW",
+      // "Rb", "Re", "Rg", "Rz",
+#endif
+  };
+  // clang-format on
+
+  current_query = queries.begin();
+}
+
+void DaikinS21::loop() {
+  using Result = DaikinSerial::Result;
+
+  switch (serial.service()) {
```

Comment:
Call the serial state machine to push along any in progress tx / rx / wait operations.
When a non-Busy result occurs, push along the command/query state machine. Usually this is via tx_next();

Which part of code is responsible for ESP_LOG could be better, I'll clean that up